File: 2024/16/ts/types/Parser.ts
import { END_TILE_REGEX, START_TILE_REGEX } from './index.ts';
import type { MazeMapInterface } from './index.ts';

export class Parser implements MazeMapInterface {
  cols: number;
  end: Position;
  maze: MazeTile[][];
  rows: number;
  start: Position;

  constructor(input: string) {
    this.maze = input.trim()
      .split('\n')
      .map((row, y) => {
        const matchesE = row.trim().match(END_TILE_REGEX);
        if (matchesE !== null) {
          this.end = { x: matchesE.index, y };
        }

        const matchesS = row.trim().match(START_TILE_REGEX);
        if (matchesS !== null) {
          this.start = { x: matchesS.index, y };
        }

        return row.trim().split('')
          .map(t => {
            return { tile: t };
          });
      });
    this.rows = this.maze.length;
    this.cols = this.maze[0].length;
  }
}

File: aoc/src/main/kotlin/com/webnologies/grkernisant/aoc/aoc2024/day16/Parser.kt
package com.webnologies.grkernisant.aoc.aoc2024.day16

class Parser : ParserInterface {
    var cols: Int = 0
    lateinit var end: Position
    var maze: List<List<MazeTile>>
    lateinit var start: Position
    var rows: Int = 0

    constructor(input: List<String>) {
        this.maze = input.mapIndexed { y, row ->
            row.trim().split("")
                .filter { it.isNotEmpty() }
                .mapIndexed { x, col ->
                    if (col == "E") this.end = Position(x, y)
                    if (col == "S") this.start = Position(x, y)

                    MazeTile(
                        Tile.of(col.toCharArray()[0]),
                        Position(x, y)
                    )
                }
        }

        this.rows = this.maze.size
        this.cols = this.maze[0].size
    }

    override fun fetchCols(): Int { return this.cols }
    override fun fetchEnd(): Position { return this.end }
    override fun fetchMazeTile(): List<List<MazeTile>> { return this.maze }
    override fun fetchRows(): Int { return this.rows }
    override fun fetchStart(): Position { return this.start }
}
